include/nbl/builtin/hlsl/bxdf/fresnel.hlsl

```diff
@@ -17,131 +18,354 @@ namespace hlsl
+    void recomputeNdotT(bool backside, scalar_type _NdotI2, scalar_type rcpOrientedEta2)
     {
-        return _N * (_NdotI * (_refract ? _rcpOrientedEta : 1.0f) + _NdotTorR) - _I * (_refract ? _rcpOrientedEta : 1.0f);
+        scalar_type NdotT2 = rcpOrientedEta2 * _NdotI2 + 1.0 - rcpOrientedEta2;
+        scalar_type absNdotT = sqrt<scalar_type>(NdotT2);
+        NdotT = ieee754::flipSign(absNdotT, backside);
     }
 
-    vector_type doReflectRefract(bool r)
+    vector_type operator()()
     {
-        const T NdotTorR = r ? computeNdotT(backside, NdotI2, rcpOrientedEta2) : NdotI;
-        return doReflectRefract(r, I, N, NdotI, NdotTorR, rcpOrientedEta);
+        recomputeNdotT(rcpOrientedEta.backside, NdotI2, rcpOrientedEta.value2);
+        return N * (NdotI * rcpOrientedEta.value + NdotT) - rcpOrientedEta.value * I;
     }
 
     vector_type I;
     vector_type N;
-    bool backside;
+    scalar_type NdotT;
     scalar_type NdotI;
     scalar_type NdotI2;
+    fresnel::OrientedEtaRcps<scalar_type> rcpOrientedEta;
+};
+
+template<typename T NBL_PRIMARY_REQUIRES(concepts::Scalar<T>)
+struct ReflectRefract
+{
+    using this_t = ReflectRefract<T>;
+    using vector_type = vector<T, 3>;
+    using scalar_type = T;
+
+    static this_t create(bool r, NBL_CONST_REF_ARG(vector_type) I, NBL_CONST_REF_ARG(vector_type) N, scalar_type NdotI, scalar_type NdotTorR, scalar_type rcpOrientedEta)
+    {
+        this_t retval;
+        retval.I = I;
+        retval.N = N;
+        retval.NdotI = NdotI;
+        retval.NdotTorR = NdotTorR;
+        retval.rcpOrientedEta = rcpOrientedEta;
+        return retval;
+    }
+
+    static this_t create(bool r, NBL_CONST_REF_ARG(Refract<scalar_type>) refract)
+    {
+        this_t retval;
+        retval.I = refract.I;
+        retval.N = refract.N;
+        retval.NdotI = refract.NdotI;
+        retval.NdotTorR = hlsl::mix(refract.NdotI, refract.NdotT, r);
+        retval.rcpOrientedEta = refract.rcpOrientedEta.value;
+        return retval;
+    }
+
+    // when you know you'll reflect
+    void recomputeNdotR()
+    {
+        refract.recomputeNdotI();
+    }
+
+    // when you know you'll refract
+    void recomputeNdotT(bool backside, scalar_type _NdotI2, scalar_type rcpOrientedEta2)
+    {
+        refract.recomputeNdotT(backside, _NdotI2, rcpOrientedEta2);
+    }
+
+    vector_type operator()(const bool doRefract)
+    {
+        return N * (NdotI * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract)) + NdotTorR) - I * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract));
+    }
+
+    Refract<scalar_type> refract;
+    vector_type I;
+    vector_type N;
+    scalar_type NdotI;
+    scalar_type NdotTorR;
     scalar_type rcpOrientedEta;
-    scalar_type rcpOrientedEta2;
 };
 
+
+namespace fresnel
+{
+
+template<typename T NBL_PRIMARY_REQUIRES(is_scalar_v<T> || is_vector_v<T>)
+struct Schlick
+{
+    using scalar_type = typename vector_traits<T>::scalar_type;
```

Comment:
concept needs to only take `FloatingPointLikeVectorial`, a scalar will fail `vector_traits` for your `scalar_type`